README.md
[![npm Version](https://img.shields.io/npm/v/@cabinfo.eu/veml6030.svg)](https://www.npmjs.com/package/@cabinfo.eu/veml6030)
[![Downloads Per Month](https://img.shields.io/npm/dm/@cabinfo.eu/veml6030.svg)](https://www.npmjs.com/package/@cabinfo.eu/veml6030)
![GitHub last commit](https://img.shields.io/github/last-commit/alaindeseine/veml6030.svg)
![GitHub top language](https://img.shields.io/github/languages/top/alaindeseine/veml6030.svg)
![GitHub package.json version](https://img.shields.io/github/package-json/v/alaindeseine/veml6030.svg)
![NPM](https://img.shields.io/npm/l/@cabinfo.eu/veml6030.svg)

# veml6030

Node.js I2C driver for the VISHAY VEML6030 ambient light sensor on Linux boards like the Raspberry Pi or BeagleBone.

Supports Node.js versions 10, 12, 14, 15 and 16.

VEML6030 chipset datasheet : https://www.vishay.com/docs/84366/veml6030.pdf

VEML6030 chipset application notes : https://www.vishay.com/docs/84367/designingveml6030.pdf?ICID=I-CT-TECH-RES-CLA-SEP_21-0


## Contents

 * [Features](#features)
 * [Installation](#installation)
 * [Usage](#usage)
 * [VEML6030 Class methods](#veml6030-class-methods)
 * [Related Packages](#related-packages)
 * [Roadmap](#Roadmap)

### Features

 * Ambient light sensing
 * Autocalibration mode
 * Manual mode
 * Promise based asynchronous call methods

### Installation

```
npm install --save veml6030
```

### Usage

#### Report the ambient light with auto calibration mode example.
```js
const VEML6030    = require('@cabinfo.eu/veml6030');

const veml6030 = new VEML6030({debug: true});

veml6030.init()
.then(() => {
    veml6030.readSensorData(true)
    .then(datas => console.log('%o', datas))
    .catch(error => console.log(error));
})
.catch(error => console.log(error));
```

Sample output:
```js
{
  rawValue: 144,
  rawLuxValue: 66.3552
  luxValue: 66.8638,
  useCorrectionFormula: true
  gain: 0.25,
  integrationTime: 50,
  autocalibrate: false,
  retry: 1,
  overflow: false
}
```


#### Report the ambient light with manual configuration example.

```js
const VEML6030    = require('@cabinfo.eu/veml6030');

const options = {
    debug: true,
    gain: 0.25,
    integrationTime: 50
};

const veml6030 = new VEML6030(options);

veml6030.init()
.then(() => {
    veml6030.readSensorData(false)
    .then(datas => console.log('Datas readed: %o', datas))
    .catch(error => console.log(error));
})
.catch(error => console.log(error));
```


Sample output:
```js
{
  rawValue: 144,
  rawLuxValue: 66.3552
  luxValue: 66.8638,
  useCorrectionFormula: true
  gain: 0.25,
  integrationTime: 50,
  autocalibrate: false,
  retry: 1,
  overflow: false
}
```

### VEML6030 class methods

- [Constructor options](#constructor-options)
- [VEML6030 methods](#veml6030-methods)
- [VEML6030 constants](#veml6030-constants)

#### Constructor options

VEML6030 class constructor accept an optionnal options object. 

None of theses options are mandatory, so you can invoke VEML6030 constructor without any parameters. In this case it will use options default values.

**Options**:

|Option name|Description|Default Value|
|:-----|:-----|:-----|
|debug|If debug is set to true, VEML6030 class will print to console debug information|false|
|i2cAddress|I2C address (in hex) of VEML6030 chipset|0x48|
|gain|Fix the gain the chipset should use for reading (use it for manual calibration mode). Authorized value are 0.125, 0.25, 1 or 2. |1|
|integrationTime|Fix the integration time (in ms) the chipset should use for reading (use it for manual calibration mode). Authorized values are 25, 50, 100, 200, 400 or 800.|100|
|ALSPersistenceProtectNumber|Leave this option to default value|1|
|ALSInterruptEnableSetting|Leave this option to default value|false|
|ALSShutDownSetting|Leave this option to default value|false|
|i2cBusNumber|I2C bus number. I most case leave this value to default value|1|

#### VEML6030 methods

- [init()](#init)
- [readSensorData()](#readSensorData)


##### init()
Returns a Promise that will be resolved with no arguments once the initial configuration has been wrote to the VEML6030 chipset, or will be rejected if an error occurs.

Configuration values are default ones if you invoke constructor without any options object. If you set some optionnal parameters in constructor call or if you invoke one or more parameter methods, it will be the current values that are send to chipset.

Once init resolve you can use the read readSensorData() method.
##### readSensorData()
Returns a Promise that will be resolved with an object once the VEML6030 chipset return the readed value, or will be rejected if an error occurs.

readSensorData accept one boolean parameter to indicat if you want an autocalibrating measure (prefered) or a raw measure with the current reading options values (gain, integration time, etc.)

If autocalibration parameter is set to true this method will adjust the gain and the integration time of the chipset according to VISHAY recommandations. 

If you set the parameter to false a simple reading is done with the parameters you set. This mode is intended to permit you to implement your own calibration method. 

We recommend using autocalibration once you don't need to implement your own measures scheme.

Object properties returned when promise resolve: 

* **rawValue**: The raw value readed from chipset
* **rawLuxValue**: The optimized raw illumination value calculated from raw value (in lux) according to VISHAY recommandations
* **luxValue**: The illumination value (in lux). If `rawLuxValue` is <= 100 `luxValue` and `rawLuxValue` are identical. If `rawLuxValue` is > 100 a correction formula is apply to `luxValue` according to VISHAY recommandations
* **useCorrectionFormula**: Indicate if a correction formula is applied to luxValue (true) or not (false)
* **gain**: The gain value used for reading
* **integrationTime**: the integration time (in ms) used for reading
* **autocalibrate**: Inditate if in autocalibration mode (true) or not (false)
* **retry**: Number of measures done. If autocalibration mode this number is > 1 due to gain and integration time adjustments.
* **overflow**: Indicate if an overflow is detected (true) or not (false)


Sample object: 

```js
{
  rawValue: 144,
  rawLuxValue: 66.3552
  luxValue: 66.8638,
  useCorrectionFormula: true
  gain: 0.25,
  integrationTime: 50,
  autocalibrate: false,
  retry: 1,
  overflow: false
}
```


#### VEML6030 constants

VEML6030 package publish folowing constants. You can use it when implementing your own implementation measures.

All theses constants are static, so you can use it like this:

```js
const VEML6030 = require('@cabinfo.eu/veml6030');
console.log('Read command is: %o', VEML6030.ALS_READ_REGISTER);
```

- **ALS_SETTING_REGISTER**: Send configuration command code. This constant value is 0x00.
- **ALS_WH_REGISTER**: Send an high value for threshold. This constant value is 0x01. This command is currently not used in this package.
- **ALS_WL_REGISTER**: Send an low value for threshold. This constant value is 0x02. This command is currently not used in this package.
- **ALS_POWER_SAVE_REGISTER**: Power saving command. This constant value is 0x03. This command is currently not used in this package.
- **ALS_READ_REGISTER**: Read command for ALS channel. This constant value is 0x04.
- **WHITE_READ_REGISTER**: Read command for white channel. This constant value is 0x05. This constant value is 0x03. This command is currently not used in this package.
- **ALS_INT_REGISTER**: Interrupt status command (use to detect low or high threshold). This constant value is 0x06. This command is currently not used in this package.
- **READ_BASE_RESOLUTION**: . This constant value is 0.0036.


### Related Packages

- [i2c-bus](https://github.com/fivdi/i2c-bus) - I2C serial bus access

### Roadmap

* Implement shutdown/powerup during autocalibration according to Vishay documents.
* Implement white channel reading.
* Implement low and high threshold and interupt status reading.


HISTORY.md
1.0.4 - Sep 29 2022
===================

  * Change decrease resolution criteria
  * Apply correction formula to readvalues > 100 
  * README.md update

1.0.3 - Sep 29 2022
===================

  * Adding manual example
  * README.md update

1.0.2 - Sep 28 2022
===================

  * Change package to a scoped one
  * README.md update

1.0.1 - Sep 28 2022
===================

  * README.md update

1.0.0 - Sep 28 2022
===================

  * initial release


test/read.js
'use strict';

const VEML6030 = require('../');

const veml6030 = new VEML6030({debug: true});

veml6030.init()
.then(() => {
    veml6030.readSensorData(true)
    .then(datas => console.log('Datas readed: %o', datas))
    .catch(error => console.log(error));
})
.catch(error => console.log(error));


index.js
'use strict';

class VEML6030 {
    static ALS_SETTING_REGISTER = 0x00;
    static ALS_WH_REGISTER = 0x01;
    static ALS_WL_REGISTER = 0x02;
    static ALS_POWER_SAVE_REGISTER = 0x03;
    static ALS_READ_REGISTER = 0x04;
    static WHITE_READ_REGISTER = 0x05;
    static ALS_INT_REGISTER = 0x06;
    static READ_BASE_RESOLUTION = 0.0036;


    constructor(options) {
        const i2c = require('i2c-bus');

        this.debug                          = (options && options.hasOwnProperty('debug')) ? options.debug : false;
        this.VEML6030_ADDR                  = (options && options.hasOwnProperty('i2cAddress')) ? options.i2cAddress : 0x48;
        this.gain                           = (options && options.hasOwnProperty('gain')) ? this.setGain(options.gain) : 0x00;
        this.integrationTime                = (options && options.hasOwnProperty('integrationTime')) ? this.setIntegrationTime(options.integrationTime) : 0x00;
        this.ALSPersistenceProtectNumber    = (options && options.hasOwnProperty('ALSPersistenceProtectNumber')) ? this.setALSPersistenceProtectNumber(options.ALSPersistenceProtectNumber) : 0x00;
        this.ALSInterruptEnableSetting      = (options && options.hasOwnProperty('ALSInterruptEnableSetting')) ? this.setALSInterruptEnable(options.ALSInterruptEnableSetting) : 0x00;
        this.ALSShutDownSetting             = (options && options.hasOwnProperty('ALSShutDownSetting')) ? this.setALSShutdown(options.ALSShutDownSetting) : 0x00;

        if (this.debug){
            console.debug('Constructor: VEML6030_ADDR: %o', this.VEML6030_ADDR);
            console.debug('Constructor: gain: %o', this.getGain(this.gain));
            console.debug('Constructor: integrationTime: %o', this.getIntegrationTime(this.integrationTime));
            console.debug('Constructor: ALSPersistenceProtectNumber: %o', this.ALSPersistenceProtectNumber);
            console.debug('Constructor: ALSInterruptEnableSetting: %o', this.ALSInterruptEnableSetting);
            console.debug('Constructor: ALSShutDownSetting: %o', this.ALSShutDownSetting);    
        }

        this.i2cBus = i2c.openSync((options && options.hasOwnProperty('i2cBusNumber')) ? options.i2cBusNumber : 1);
    } // eo constructor


    init() {
        return new Promise((resolve, reject) => {
            this.writeConfiguration()
            .then(() => {
                resolve();
            })
            .catch( error => reject(error));
        });
    } // eo init method


    calculateLuxLevel(rawValue){
        let integrationFactor;
        let gainFactor;

        switch(this.gain){
            case 0x01 : {
                gainFactor = 1;
                break;
            }
            case 0x00 : {
                gainFactor = 2;
                break;
            }
            case 0x03 : {
                gainFactor = 8;
                break;
            }
            case 0x02 : {
                gainFactor = 16;
                break;
            }
        }

        switch(this.integrationTime){
            case 0x03 : {
                integrationFactor = 1;
                break;
            }
            case 0x02 : {
                integrationFactor = 2;
                break;
            }
            case 0x01 : {
                integrationFactor = 4;
                break;
            }
            case 0x00 : {
                integrationFactor = 8;
                break;
            }
            case 0x08 : {
                integrationFactor = 16;
                break;
            }
            case 0x0c : {
                integrationFactor = 32;
                break;
            }
        }

        return rawValue * VEML6030.READ_BASE_RESOLUTION * gainFactor * integrationFactor;

    } // eo calculateLuxLevel method


    applyCorrectionFormula(rawValue){
        return (6.0135e-13 * (rawValue ** 4)) - (9.3924e-09 * (rawValue ** 3)) + (8.1488e-05 * (rawValue ** 2)) + (1.0023e-00 * rawValue);
    } // eo applyCorrectionFormula method


    setGain(gain){
        switch(gain){
            case 0.125 : {
                this.gain = 0x02;
                break;
            }
            case 0.25 : {
                this.gain = 0x03;
                break;
            }
            case 1.0 : {
                this.gain = 0x00;
                break;
            }
            case 2.0 : {
                this.gain = 0x01;
                break;
            }
            default : {
                this.gain = 0x00;
            }
        }

        return this.gain;
    } // eo setGain method


   getGain(){
        switch(this.gain){
            case 0x03 : {
                return 0.25;
            }
            case 0x02 : {
                return 0.125;
            }
            case 0x01 : {
                return 2.0;
            }
            case 0x00 : {
                return 1.0;
            }
        }
    } // eo getGain method


    setIntegrationTime(integrationTime){
        switch(integrationTime){
            case 25 : {
                this.integrationTime = 0x0c;
                break;
            }
            case 50 : {
                this.integrationTime = 0x08;
                break;
            }
            case 100 : {
                this.integrationTime = 0x00;
                break;
            }
            case 200 : {
                this.integrationTime = 0x01;
                break;
            }
            case 400 : {
                this.integrationTime = 0x02;
                break;
            }
            case 800 : {
                this.integrationTime = 0x03;
                break;
            }
            default : {
                this.integrationTime = 0x00;
            }
        }

        return this.integrationTime;
    } // eo setIntegrationTime method


    getIntegrationTime(){
        switch(this.integrationTime){
            case 0x03 : {
                return 800;
            }
            case 0x02 : {
                return 400;
            }
            case 0x01 : {
                return 200;
            }
            case 0x00 : {
                return 100;
            }
            case 0x08 : {
                return 50;
            }
            case 0x0c : {
                return 25;
            }
        }
    } // eo getIntegrationTime method


    setALSPersistenceProtectNumber(protectNumber){
        switch(protectNumber){
            case 1 : {
                this.ALSPersistenceProtectNumber = 0x00;
                break;
            }
            case 2 : {
                this.ALSPersistenceProtectNumber = 0x01;
                break;
            }
            case 4 : {
                this.ALSPersistenceProtectNumber = 0x02;
                break;
            }
            case 8 : {
                this.ALSPersistenceProtectNumber = 0x03;
                break;
            }
            default : {
                this.ALSPersistenceProtectNumber = 0x00;
            }
        }

        return this.ALSPersistenceProtectNumber;
    } // eo setALSPersistenceProtectNumber method


    setALSInterruptEnable(protectNumber){
        switch(protectNumber){
            case false : {
                this.ALSInterruptEnableSetting = 0x00;
                break;
            }
            case true : {
                this.ALSInterruptEnableSetting = 0x01;
                break;
            }
            default : {
                this.ALSInterruptEnableSetting = 0x00;
            }
        }

        return this.ALSInterruptEnableSetting;
    } // eo setALSPersistenceProtectNumber method


    setALSShutdown(protectNumber){
        switch(protectNumber){
            case false : {
                this.ALSShutDownSetting = 0x00;
                break;
            }
            case true : {
                this.ALSShutDownSetting = 0x01;
                break;
            }
            default : {
                this.ALSShutDownSetting = 0x00;
            }
        }

        return this.ALSShutDownSetting;
    } // eo setALSShutdown method


    writeConfiguration(){
        return new Promise((resolve, reject) => {
            let command = (this.gain << 11) + (this.integrationTime << 6) + (this.ALSPersistenceProtectNumber << 4) + (this.ALSInterruptEnableSetting << 1) + (this.ALSShutDownSetting << 0);

            if (this.debug){
                console.log('Configuration: %o', command.toString(2));
            }

            this.i2cBus.writeWord(this.VEML6030_ADDR, VEML6030.ALS_SETTING_REGISTER, command, (error) => {
                if (error) {
                    return reject(error);
                } 
                else {
                    if(this.debug){
                        console.debug('writeConfiguration: configuration saved');
                    }

                    return resolve(true);
                }
            });
        });
    } // eo writeConfiguration method


    async writeConfigurationSync() {
        let command = (this.gain << 11) + (this.integrationTime << 6) + (this.ALSPersistenceProtectNumber << 4) + (this.ALSInterruptEnableSetting << 1) + (this.ALSShutDownSetting << 0);

        if (this.debug){
            console.log('Configuration: %o', command.toString(2));
        }

        this.i2cBus.writeWordSync(this.VEML6030_ADDR, VEML6030.ALS_SETTING_REGISTER, command);

        if(this.debug){
            console.debug('writeConfigurationSync: configuration saved');
        }
    } // eo writeConfigurationSync method


    readSensorData(autoCalibrate = false) {
        return new Promise((resolve, reject) => {
            if (! autoCalibrate){
                // No auto calibrate, make a simple measure.
                this.i2cBus.readWord(this.VEML6030_ADDR, VEML6030.ALS_READ_REGISTER, (error, readValue) => {
                    if(error) {
                        return reject(error);
                    }

                    let rawLuxValue             = this.calculateLuxLevel(readValue);
                    let luxValue                = rawLuxValue;
                    let useCorrectionFormula    = false;
    
                    if (readValue > 100){
                        luxValue    = this.applyCorrectionFormula(rawLuxValue);
                        useCorrectionFormula    = true;
                    }
    
                    let overflow    = this.checkForOverflow(luxValue);
    
                    return resolve({
                        rawValue : readValue,
                        rawLuxValue: rawLuxValue,
                        luxValue : luxValue,
                        useCorrectionFormula: useCorrectionFormula,
                        gain : this.getGain(this.gain),
                        integrationTime : this.getIntegrationTime(this.integrationTime),
                        autocalibrate: false,
                        retry: 1,
                        overflow: overflow
                    });
                });                
            }
            else {
                // Auto calibrate.
                let readValue = 0;
                let timeoutCount = 25;
                let count = 0;
                let hasValue = false;
                // Set initial gain and integration time.
                this.gain = this.setGain(0.125);
                this.integrationTime = this.setIntegrationTime(100);
                this.writeConfigurationSync();

                while (! hasValue){
                    // Check if we should go to timeout.
                    if (count == timeoutCount){
                        hasValue = true;
                        return reject("Reading timeout");
                    }

                    // read value from I2C
                    readValue = this.i2cBus.readWordSync(this.VEML6030_ADDR, VEML6030.ALS_READ_REGISTER);

                    // Check if readed value is lower than 100.
                    if (readValue <= 100) {
                        // Readed value is lower than 100, increase resolution and make measure again.
                        if (this.debug){
                            console.debug('readValue: %o', readValue);
                        }

                        this.increaseResolution();
                        this.writeConfigurationSync();
                    }
                    else {
                        // Readed value is greater than 100, check if > 10 000.
                        if (readValue > 10000){
                            // Decrease resolution.
                            this.decreaseResolution();
                            this.writeConfigurationSync();
                        }
                        else {
                            // readValue i > 100 and < 10000, reading is ok.
                            hasValue = true;
                        }
                    }

                    count++;
                }

                let rawLuxValue             = this.calculateLuxLevel(readValue);
                let luxValue                = rawLuxValue;
                let useCorrectionFormula    = false;

                if (readValue > 100){
                    luxValue    = this.applyCorrectionFormula(rawLuxValue);
                    useCorrectionFormula    = true;
                }

                let overflow    = this.checkForOverflow(luxValue);

                return resolve({
                    rawValue : readValue,
                    rawLuxValue: rawLuxValue,
                    luxValue : luxValue,
                    useCorrectionFormula: useCorrectionFormula,
                    gain : this.getGain(this.gain),
                    integrationTime : this.getIntegrationTime(this.integrationTime),
                    autocalibrate: true,
                    retry: count,
                    overflow: overflow
                });
            }
        });
    } // eo readSensorData method


    checkForOverflow(readValue){
        let isOverflow = false;

        if ( this.gain == 0x01 && this.integrationTime == 0x03 && readValue >= 236){
            isOverflow = true;
        }

        if ( this.gain == 0x01 && this.integrationTime == 0x02 && readValue >= 472){
            isOverflow = true;
        }

        if ( this.gain == 0x01 && this.integrationTime == 0x01 && readValue >= 944){
            isOverflow = true;
        }

        if ( this.gain == 0x01 && this.integrationTime == 0x00 && readValue >= 1887){
            isOverflow = true;
        }

        if ( this.gain == 0x01 && this.integrationTime == 0x08 && readValue >= 3775){
            isOverflow = true;
        }

        if ( this.gain == 0x01 && this.integrationTime == 0x0c && readValue >= 7550){
            isOverflow = true;
        }

        if ( this.gain == 0x00 && this.integrationTime == 0x03 && readValue >= 472){
            isOverflow = true;
        }

        if ( this.gain == 0x00 && this.integrationTime == 0x02 && readValue >= 944){
            isOverflow = true;
        }

        if ( this.gain == 0x00 && this.integrationTime == 0x01 && readValue >= 1887){
            isOverflow = true;
        }

        if ( this.gain == 0x00 && this.integrationTime == 0x00 && readValue >= 3775){
            isOverflow = true;
        }

        if ( this.gain == 0x00 && this.integrationTime == 0x08 && readValue >= 7550){
            isOverflow = true;
        }

        if ( this.gain == 0x00 && this.integrationTime == 0x0c && readValue >= 15099){
            isOverflow = true;
        }

        if ( this.gain == 0x03 && this.integrationTime == 0x03 && readValue >= 1887){
            isOverflow = true;
        }

        if ( this.gain == 0x03 && this.integrationTime == 0x02 && readValue >= 3775){
            isOverflow = true;
        }

        if ( this.gain == 0x03 && this.integrationTime == 0x01 && readValue >= 7550){
            isOverflow = true;
        }

        if ( this.gain == 0x03 && this.integrationTime == 0x00 && readValue >= 15099){
            isOverflow = true;
        }

        if ( this.gain == 0x03 && this.integrationTime == 0x08 && readValue >= 30199){
            isOverflow = true;
        }

        if ( this.gain == 0x03 && this.integrationTime == 0x0c && readValue >= 60398){
            isOverflow = true;
        }

        if ( this.gain == 0x02 && this.integrationTime == 0x03 && readValue >= 3775){
            isOverflow = true;
        }

        if ( this.gain == 0x02 && this.integrationTime == 0x02 && readValue >= 7550){
            isOverflow = true;
        }

        if ( this.gain == 0x02 && this.integrationTime == 0x01 && readValue >= 15099){
            isOverflow = true;
        }

        if ( this.gain == 0x02 && this.integrationTime == 0x00 && readValue >= 30199){
            isOverflow = true;
        }

        if ( this.gain == 0x02 && this.integrationTime == 0x08 && readValue >= 60398){
            isOverflow = true;
        }

        if ( this.gain == 0x02 && this.integrationTime == 0x0c && readValue >= 120796){
            isOverflow = true;
        }

        return isOverflow;
    }

    increaseResolution(){
        let changed = false;

        if (this.debug){
            console.debug('increaseResolution (before): this.gain: %o', this.gain);
            console.debug('increaseResolution (before): this.integrationTime: %o', this.integrationTime);
        }

        if (this.gain == 0x02 && ! changed){
            // gain = 0.125 set to 0.25 
            this.setGain(0.25);
            changed = true;
        }

        if (this.gain == 0x03 && ! changed){
            // gain = 0.25 set to 1.0 
            this.setGain(1.0);
            changed = true;
        }

        if (this.gain == 0x00 && ! changed){
            // gain = 1.0 set to 2.0 
            this.setGain(2.0);
            changed = true;
        }

        if (this.gain == 0x01 && ! changed){
            // Maximum gain, set Integration time
            if (this.integrationTime == 0x0c && ! changed){
                // Integration time = 25 ms set to 50 ms
                this.setIntegrationTime(50);
                changed = true;
            }

            if (this.integrationTime == 0x08 && ! changed){
                // Integration time = 50 ms set to 100 ms
                this.setIntegrationTime(100);
                changed = true;
            }

            if (this.integrationTime == 0x00 && ! changed){
                // Integration time = 100 ms set to 200 ms
                this.setIntegrationTime(200);
                changed = true;
            }

            if (this.integrationTime == 0x01 && ! changed){
                // Integration time = 200 ms set to 400 ms
                this.setIntegrationTime(400);
                changed = true;
            }

            if (this.integrationTime == 0x02 && ! changed){
                // Integration time = 400 ms set to 800 ms
                this.setIntegrationTime(800);
                // changed = true;
            }
        }

        if (this.debug){
            console.debug('increaseResolution (after): this.gain: %o', this.gain);
            console.debug('increaseResolution (after): this.integrationTime: %o', this.integrationTime);
        }
    } // eo increaseResolution method


    decreaseResolution(){
        let changed = false;

        if (this.integrationTime == 0x03 && ! changed){
            // Integration time = 800 ms set to 400 ms
            this.setIntegrationTime(400);
            changed = true;
        }

        if (this.integrationTime == 0x02 && ! changed){
            // Integration time = 400 ms set to 200 ms
            this.setIntegrationTime(200);
            changed = true;
        }

        if (this.integrationTime == 0x01 && ! changed){
            // Integration time = 200 ms set to 100 ms
            this.setIntegrationTime(100);
            changed = true;
        }

        if (this.integrationTime == 0x00 && ! changed){
            // Integration time = 100 ms set to 50 ms
            this.setIntegrationTime(50);
            changed = true;
        }

        if (this.integrationTime == 0x08 && ! changed){
            // Integration time = 50 ms set to 25 ms
            this.setIntegrationTime(25);
            changed = true;
        }

        if (this.integrationTime == 0x0c && ! changed){
            // Minimum Integration time (25 ms), decrease gain
            if (this.gain == 0x01 && ! changed){
                // gain = 2.0 set to 1.0 
                this.setGain(1.0);
                changed = true;
            }

            if (this.gain == 0x00 && ! changed){
                // gain = 1.0 set to 0.25 
                this.setGain(0.25);
                changed = true;
            }

            if (this.gain == 0x03 && ! changed){
                // gain = 0.25 set to 0.125 
                this.setGain(0.125);
                // changed = true;
            }
        }
    } // eo decreaseResolution method

} // eo VEML6030 class

module.exports = VEML6030;